Dalamud/Game/Gui/NamePlate/NamePlateGui.cs

```diff
@@ -0,0 +1,302 @@
+            {
+                handler.ResetState();
+                if (handler.IsUpdating)
+                    udpatedHandlers.Add(handler);
+            }
+
+            if (this.OnDataUpdate is not null)
+            {
+                this.OnDataUpdate?.Invoke(this.context, activeHandlers);
+                this.OnNamePlateUpdate?.Invoke(this.context, udpatedHandlers);
+                if (this.context.HasParts)
+                    this.ApplyBuilders(activeHandlers);
+            }
+            else if (udpatedHandlers.Count != 0)
+            {
+                var changedHandlersSpan = udpatedHandlers.ToArray().AsSpan();
+                this.OnNamePlateUpdate?.Invoke(this.context, udpatedHandlers);
+                if (this.context.HasParts)
+                    this.ApplyBuilders(changedHandlersSpan);
+            }
+        }
+    }
+
+    private void ApplyBuilders(Span<NamePlateUpdateHandler> handlers)
+    {
+        foreach (var handler in handlers)
+        {
+            if (handler.PartsContainer is { } container)
+            {
+                container.ApplyBuilders(handler);
+            }
+        }
+    }
+}
+
+/// <summary>
+/// Plugin-scoped version of a AddonEventManager service.
+/// </summary>
+[PluginInterface]
+[ServiceManager.ScopedService]
+#pragma warning disable SA1015
+[ResolveVia<INamePlateGui>]
+#pragma warning restore SA1015
+internal class NamePlateGuiPluginScoped : IInternalDisposableService, INamePlateGui
+{
+    [ServiceManager.ServiceDependency]
+    private readonly NamePlateGui parentService = Service<NamePlateGui>.Get();
+
+    /// <inheritdoc/>
+    public event INamePlateGui.OnPlateUpdateDelegate? OnNamePlateUpdate
+    {
+        add
+        {
+            if (this.OnNamePlateUpdateScoped == null)
+                this.parentService.OnNamePlateUpdate += this.OnNamePlateUpdateForward;
+            this.OnNamePlateUpdateScoped += value;
+        }
+
+        remove
+        {
+            this.OnNamePlateUpdateScoped -= value;
+            if (this.OnNamePlateUpdateScoped == null)
+                this.parentService.OnNamePlateUpdate -= this.OnNamePlateUpdateForward;
+        }
+    }
+
+    /// <inheritdoc/>
+    public event INamePlateGui.OnPlateUpdateDelegate? OnDataUpdate
+    {
+        add
+        {
+            if (this.OnDataUpdateScoped == null)
+                this.parentService.OnDataUpdate += this.OnDataUpdateForward;
+            this.OnDataUpdateScoped += value;
+        }
+
+        remove
+        {
+            this.OnDataUpdateScoped -= value;
+            if (this.OnDataUpdateScoped == null)
+                this.parentService.OnDataUpdate -= this.OnDataUpdateForward;
+        }
+    }
+
+    private event INamePlateGui.OnPlateUpdateDelegate? OnNamePlateUpdateScoped;
+
+    private event INamePlateGui.OnPlateUpdateDelegate? OnDataUpdateScoped;
+
+    /// <inheritdoc/>
+    public void RequestRedraw()
+    {
+        this.parentService.RequestRedraw();
```

Comment:
Does it matter if multiple plugins call this very frequently/multiple times in the same frame? I assume some game code reads that flag set in the number array and re-renders the nameplates?

Reply:
Right. The game reads the flag we set which tells it to do a full update, and we poke the array data to cause an update on the next frame. Basically it's not a problem to call every frame if you're dragging a slider around in a settings window or something, and multiple plugins can call it in the same frame without any issue. But if a plugin decided to call it every frame during gameplay as a matter of course that would not be good for performance.